### ISSUE TITLE
Unexpected behaviour in validation results

### ISSUE THREAD

**Issue by cookielee77** [neural-nuts/image-caption-generator#11]
the 10000 is so large. Why we should train the Flickr30k so many epochs?

Also, I trained the graph almost 200 epochs, but the results look still not favorable: 
a man in a black shirt and black pants is sitting on a bench . 
a man is sitting on a bench with a red shirt and a black shirt . 
a man is sitting on a bench with a man in a black shirt . 
a man in a black shirt and black pants is sitting on a bench . 
a man in a black shirt and a black shirt is sitting on a bench . 
Although they came from difference images. 

Any clue? I used one TitanX for training, which follows your instructions. 

Thank you very much!

**Comment by metalaman** [comment]
10000 epochs is an upper limit. We haven't trained the model for that long. It's good practice if we keep track of the training and validation accuracy after every ~20-50 epochs to see if the model is actually improving. Once the accuracy of validation doesn't change much, we stop the training. This is called early stopping and is used to avoid overfitting on data. 
Try using the above approach to find a good fit for the model. Also, since you have a powerful GPU, try training on MSCOCO dataset.

**Comment by cookielee77** [comment]
Oh, I see. So this issue is caused by overfitting, right?  How many epochs did you usually train the model until it reachs a good performance? 

Also, the image captioning task do not have accuracy metric. Did you use something else to monitor the training? (For example, BLUE score? )

Thanks!

**Comment by pranay360** [comment]
We trained for about 50 epochs, we didn't have a good GPU.
check eval.py for validation.

**Comment by dust0x** [comment]
@cookielee77 could you share an update on the model performance with early stopping?

**Comment by cookielee77** [comment]
Hi, dust0x, sorry for this late reply. I was attending NIPS2017 last week and just came back. I'll test the early stopping method and give your feedback right away. Thanks!

**Comment by cookielee77** [comment]
Hi, the model generates reasonable captions. However, I still don't know when the model has best performance. I used 1000 images as the validation set and monitored the training by using BLEU metrics. But I found the BLEU scores don't improve much during each epoch, they are almost the same around:

Mean Sentence-Level BLEU 1: 0.583397, BLEU 2: 0.419192, BLEU 3: 0.414963, BLEU 4: 0.450221

I used the generated sentence against the five ground truths to calculate the sentence-level scores.

**Comment by metalaman** [comment]
Hi, are you using [this](https://github.com/tylin/coco-caption) to calculate the scores. We used this and so did O. Vinays et. al.(Show and Tell).

**Comment by cookielee77** [comment]
Yes, I used that to calculate the BLEU, CIDEr, etc. **The email with log file** I sent to you yesterday used the pycocoevalcap to calculate scores. However, the model has best validation scores in 10 epochs, not the 60 epochs (on the flicker30k dataset), and the best scores on validation I got are: 

BLEU 1: 0.595085, BLEU 2: 0.381093, BLEU 3: 0.247346, BLEU 4: 0.162187
METERO: 0.177543, ROUGE_L 0.445139, CIDEr: 0.298163

The BLEU1 score has a gap with the socre which the paper reports, which is 0.66. In 10 epochs, the average training loss is still 2.028, not very low. 

For flicker30k, I used the first 1000 images as the validation set. Other parameters are the same as your github's. I also changed the num_hidden from 256 to 512, but there is no significant improvement.

**Comment by metalaman** [comment]
>However, the model has best validation scores in 10 epochs, not the 60 epochs (on the flicker30k dataset)

Not sure why such a thing should happen.
We evaluated our model on COCO and got the following scores:
BLEU 1: 65.0 BLEU 2: 46.9 BLEU 3: 36.2 BLEU 4: 22.7 ROUGE-L: 48.0 METEOR: 22.0 CIDEr: 74.7

>The BLEU1 score has a gap with the socre which the paper reports, which is 0.66.

The paper uses ensemble of model and beam search, but still shouldn't result in huge score jumps.
I'd suggest if you have the resources try training on COCO.

**Comment by cookielee77** [comment]
Did you evaluate the model on flicker30k dataset? Since your codes on github are based on flicker30k dataset, so I tested it by following your instruction (except the validation step), but received unfavorable scores. What's the scores you got in the flicker30k dataset?

Yeah, your scores on COCO look reasonable. When you train the model on COCO, did you change the parameters of model in the github? (For example, hidden_num: 256 -> 512, and droput: disable -> enable, according to the paper) BTW, do you mind sending the codes you receive such scores on COCO to me? I definitely will cite and acknowledge your codes in our paper if it works. Thanks!

**Comment by pranay360** [comment]
@cookielee77 

**The hyperparameters (on GitHub) are currently calibrated for _MSCOCO_ dataset**, that is why you are getting unsatisfactory results for _flickr30K_.
We had to shift to _MSCOCO_ because it is better and bigger than _flick30K_, and I forgot the update dataset details in the ReadMe. I'll update the code and ReadMe and soon as possible to avoid further confusion.
Meanwhile, if you wish to run this code for _MSCOCO_ dataset, please have a look at [_utils/PreProcessCOCOCaptions.ipynb_](https://github.com/neural-nuts/image-caption-generator/blob/master/utils/PreProcessCOCOCaptions.ipynb) to make _COCO_ image-captions mapped data compatible with the model code(similar to _flickr30K_ image-captions map in _result_X.token_), then generate features for COCO images using _convfeatures.py_ then use these features and captions to train the model for COCO(you'll have to change the input file names in _main.py_).

I hope this resolves this issue.

**Comment by jishan900** [comment]
Hello all,
when i want to try run this project always show this error : 
---------------------------------------------------------------------------------------------------------------------
jishan@jishan-HP-ProBook-450-G3:~/Downloads/project$ python main.py --mode trainRequired Files not present. Regenerating Data.
Loading Caption Data Dataset/results_20130124.token
Traceback (most recent call last):
  File "main.py", line 55, in <module>
    config.word_threshold, config.max_len, caption_file, feature_file)
  File "/home/jishan/Downloads/project/utils/data_util.py", line 116, in generate_captions
    captions = [caps.replace('\n', '').split('\t')[1] for caps in data]
IndexError: list index out of range
----------------------------------------------------------------------------------------------------
how can i fix this error?please help me bro.please tell me what's the solution.I already run python convfeatures.py .But when i want to try run main.py then show this problem.Please help me.
Thanks in advance

**Comment by cookielee77** [comment]
Thanks! I will wait for your final codes and parameters for both flick30k and coco dataset. 
Actually, I already trained the codes on COCO dataset. I built my own customized prepare_dataset.py for COCO(including extraction features and process the captions index), the basic model is based on your codes. However, the model will collapse after the first epoch. Not sure what happen, I will update once I found something.
